display_list/display_list_builder.cc

```diff
@@ -413,19 +414,29 @@ void DisplayListBuilder::setAttributesFromPaint(
   }
 }
 
+void DisplayListBuilder::checkForDeferredSave() {
+  if (current_layer_->has_deferred_save_op_) {
+    Push<SaveOp>(0, 1);
+    current_layer_->has_deferred_save_op_ = false;
+  }
+}
+
 void DisplayListBuilder::save() {
-  Push<SaveOp>(0, 1);
   layer_stack_.emplace_back(current_layer_);
   current_layer_ = &layer_stack_.back();
+  current_layer_->has_deferred_save_op_ = true;
 }
+
 void DisplayListBuilder::restore() {
+  if (!current_layer_->has_deferred_save_op_) {
```

Comment:
This should be done inside the layer_stack.size() check. If the layer stack is down to 1 element we never want to emit a restore because there can be no outstanding save calls.